File: src/main/java/com/github/dylmeadows/eontimer/model/timers/EnhancedEntralinkTimer.java
package com.github.dylmeadows.eontimer.model.timers;

import com.github.dylmeadows.eontimer.model.Console;
import com.github.dylmeadows.eontimer.reference.timer.Gen5TimerConstants;

import java.time.Duration;

public class EnhancedEntralinkTimer extends EntralinkTimer implements Gen5TimerConstants {

    private final int targetAdvances;

    private final int npcCount;

    private final int initialAdvances;

    private final int frameCalibration;

    public EnhancedEntralinkTimer(int calibration,
                                  int entralinkCalibration,
                                  int targetDelay,
                                  int targetSecond,
                                  int targetAdvances,
                                  int frameCalibration,
                                  Console console,
                                  int minimumLength) {
        this(calibration, entralinkCalibration, targetDelay, targetSecond, targetAdvances, frameCalibration, console, minimumLength, true);
    }

    public EnhancedEntralinkTimer(int calibration,
                                  int entralinkCalibration,
                                  int targetDelay,
                                  int targetSecond,
                                  int targetAdvances,
                                  int frameCalibration,
                                  Console console,
                                  int minimumLength,
                                  boolean initialize) {
        super(calibration, entralinkCalibration, targetDelay, targetSecond, console, minimumLength, false);
        this.targetAdvances = targetAdvances;
        this.npcCount = 0;
        this.initialAdvances = 0;
        this.frameCalibration = frameCalibration;

        // TODO: refactor
        /*if (initialize) initialize();*/
    }

    public static int calibrate(EnhancedEntralinkTimer timer, int result) {
        // TODO: refactor
        /*double npcRate = 1.0 / Calibrations.convertToMillis(32, timer.getConsole());
        return (int) Math.round((timer.getTargetAdvances() - result) / (ENTRALINK_FRAME_RATE + (timer.getNpcCount() * npcRate))) * 1000;*/
        return 0;
    }

    @Override
    protected Duration getStage(int stage) {
        // TODO: refactor
        /*switch (stage) {
            case 0:
            case 1:
                return super.getStage(stage);
            case 2:
                return calcFrameTime();
            default:
                return TimerConstants.NULL_TIME_SPAN;
        }*/
        return Duration.ZERO;
    }

    public int getMinutesBeforeTarget() {
        // TODO: refactor
        /*Duration duration = Duration.ofMillis(0);

        for (int i = 0; i < 2; i++) {
            duration = duration.plus(getStages().get(i));
        }

        return (int) duration.toMillis() / 60000;*/
        return 0;
    }

    private Duration calcFrameTime() {
        // TODO: refactor
        /*int advances = getTargetAdvances() - getInitialAdvances();
        double npcRate = 1.0 / Calibrations.convertToMillis(32, getConsole());
        int ms = (int) Math.round((advances) / (ENTRALINK_FRAME_RATE + (getNpcCount() * npcRate))) * 1000 + getFrameCalibration();
        return Duration.ofMillis(ms);*/
        return Duration.ZERO;
    }

    public int getTargetAdvances() {
        return targetAdvances;
    }

    public int getNpcCount() {
        return npcCount;
    }

    public int getInitialAdvances() {
        return initialAdvances;
    }

    public int getFrameCalibration() {
        return frameCalibration;
    }
}

File: src/services/timers/EnhancedEntralinkTimer.cpp
//
// Created by Dylan Meadows on 2020-03-22.
//

#include "EnhancedEntralinkTimer.h"

#include <cmath>

namespace service::timer {
    const double ENTRALINK_FRAME_RATE = 0.837148929;

    EnhancedEntralinkTimer::EnhancedEntralinkTimer(
        const EntralinkTimer *entralinkTimer)
        : entralinkTimer(entralinkTimer) {}

    const std::shared_ptr<std::vector<int>>
    EnhancedEntralinkTimer::createStages(const int targetDelay,
                                         const int targetSecond,
                                         const int targetAdvances,
                                         const int calibration,
                                         const int entralinkCalibration,
                                         const int frameCalibration) const {
        std::shared_ptr<std::vector<int>> stages =
            std::make_shared<std::vector<int>>(3);
        (*stages)[0] = createStage1(targetDelay, targetSecond, calibration);
        (*stages)[1] =
            createStage2(targetDelay, calibration, entralinkCalibration);
        (*stages)[2] = createStage3(targetAdvances, frameCalibration);
        return stages;
    }

    int EnhancedEntralinkTimer::createStage1(const int targetDelay,
                                             const int targetSecond,
                                             const int calibration) const {
        return entralinkTimer->createStage1(targetDelay, targetSecond,
                                            calibration);
    }

    int EnhancedEntralinkTimer::createStage2(
        const int targetDelay, const int calibration,
        const int entralinkCalibration) const {
        return entralinkTimer->createStage2(targetDelay, calibration,
                                            entralinkCalibration);
    }

    int EnhancedEntralinkTimer::createStage3(const int targetAdvances,
                                             const int frameCalibration) const {
        return static_cast<int>(round(targetAdvances / ENTRALINK_FRAME_RATE) *
                                    1000 +
                                frameCalibration);
    }

    int EnhancedEntralinkTimer::calibrate(int targetAdvances,
                                          int actualAdvances) const {
        return static_cast<int>((targetAdvances - actualAdvances) /
                                ENTRALINK_FRAME_RATE) *
               1000;
    }
}  // namespace service::timer
